File: control-plane/sdk/python/aosdk/gateway.py
"""Gateway client — typed tasks + streaming (issue #41).

Wraps the model-gateway dispatch surface ``POST /v1/agents/{agentId}/tasks``
(issue #16) behind typed, authenticated calls:

- :meth:`GatewayClient.dispatch` submits one task and returns a typed
  :class:`~aosdk.model.TaskResult` whose ``content`` is the schema-validated
  typed object on a served outcome (``success``/``cache_hit``) and whose
  ``outcome`` is one of the closed set otherwise — the SDK never fabricates
  content for a non-served outcome.
- :meth:`GatewayClient.stream` uses streaming semantics: it yields the
  incremental dispatch events then the terminal :class:`TaskResult`,
  mirroring the platform's streaming handler (issue #16).

Every request carries a short-lived per-tenant session token from the
:class:`~aosdk.auth.TokenSource` (env or callback) — never a hardcoded key.
"""

from __future__ import annotations

from typing import Any, Dict, Iterator, Mapping, Optional, Union

from .auth import TokenSource, verify_not_expired
from .errors import ApiError, ConfigurationError
from .model import (
    DispatchEvent,
    TaskEnvelope,
    TaskRequest,
    TaskResult,
)


class GatewayClient:
    """Typed client for the model-gateway dispatch surface."""

    def __init__(
        self,
        transport: Any,
        *,
        token_source: Optional[TokenSource] = None,
        tenant_id: Optional[str] = None,
    ) -> None:
        self._transport = transport
        self._token_source = token_source or TokenSource()
        self._tenant_id = tenant_id

    # -- auth / tenant resolution ------------------------------------------- #
    def _token_and_tenant(self) -> tuple:
        token = self._token_source.require()
        session = verify_not_expired(token)  # fail closed on a lapsed token
        tenant_id = self._tenant_id or session.tenant_id
        if not tenant_id:
            raise ConfigurationError(
                "no tenant context: pass tenant_id or use a session token with a tenantId claim"
            )
        return token, tenant_id

    # -- single dispatch ----------------------------------------------------- #
    def dispatch(
        self,
        agent_id: str,
        task_type: str,
        *,
        input_: Optional[Mapping[str, Any]] = None,
        complexity: Optional[float] = None,
        tokens: Optional[int] = None,
        request_id: Optional[str] = None,
        metadata: Optional[Mapping[str, Any]] = None,
    ) -> TaskResult:
        """Submit one task and return the typed :class:`TaskResult`.

        A served outcome carries the schema-validated typed ``content``; any
        other outcome is explicit (``denied``/``blocked``/``failed``/...) and
        carries no fabricated content.  Use :meth:`run` to raise on a
        non-served outcome.
        """
        token, tenant_id = self._token_and_tenant()
        request = TaskRequest(
            tenant_id=tenant_id,
            task_type=task_type,
            input=dict(input_ or {}),
            complexity=complexity,
            tokens=tokens,
            stream=False,
            request_id=request_id,
            metadata=dict(metadata or {}),
        )
        response = self._transport.request(
            "POST",
            f"/v1/agents/{agent_id}/tasks",
            body=request.to_wire(),
            token=token,
        )
        return self._parse_gateway_response(response, agent_id, tenant_id, task_type)

    def run(self, agent_id: str, task_type: str, **kwargs: Any) -> TaskResult:
        """Dispatch and raise :class:`TaskNotServedError` on a non-served outcome."""
        return self.dispatch(agent_id, task_type, **kwargs).ensure_served()

    # -- streaming dispatch --------------------------------------------------- #
    def stream(
        self,
        agent_id: str,
        task_type: str,
        *,
        input_: Optional[Mapping[str, Any]] = None,
        complexity: Optional[float] = None,
        tokens: Optional[int] = None,
        request_id: Optional[str] = None,
        metadata: Optional[Mapping[str, Any]] = None,
    ) -> Iterator[Union[DispatchEvent, TaskResult]]:
        """Stream one dispatch: yields incremental events, then the terminal result.

        The transport must implement the :class:`~aosdk.transport.StreamTransport`
        contract (the offline doubles do; production wiring provides an SSE
        relay adapter).
        """
        request_stream = getattr(self._transport, "request_stream", None)
        if request_stream is None:
            raise ConfigurationError(
                "transport does not support streaming; provide a StreamTransport"
            )
        token, tenant_id = self._token_and_tenant()
        request = TaskRequest(
            tenant_id=tenant_id,
            task_type=task_type,
            input=dict(input_ or {}),
            complexity=complexity,
            tokens=tokens,
            stream=True,
            request_id=request_id,
            metadata=dict(metadata or {}),
        )
        for chunk in request_stream(
            "POST",
            f"/v1/agents/{agent_id}/tasks",
            body=request.to_wire(),
            token=token,
        ):
            if "event" in chunk:
                yield DispatchEvent.from_dict(chunk["event"])
            else:
                yield self._parse_gateway_response(chunk, agent_id, tenant_id, task_type)

    # -- parsing -------------------------------------------------------------- #
    @staticmethod
    def _parse_gateway_response(
        response: Dict[str, Any], agent_id: str, tenant_id: str, task_type: str
    ) -> TaskResult:
        if not isinstance(response, Mapping):
            raise ApiError(500, "malformed_response", "gateway response was not an object")
        if "result" not in response:
            from .envelope import error_from_envelope

            raise error_from_envelope(response)
        envelope = TaskEnvelope.from_dict(response)
        result = envelope.result
        if not result.request_id and not result.tenant_id:
            # Defensive: never hand back an empty result on a 4xx/5xx error body.
            raise ApiError(
                envelope.status, "malformed_response", "gateway envelope had no result payload"
            )
        return result

File: control-plane/sdk/typescript/src/gateway.ts
/**
 * Gateway client (TypeScript) — typed tasks + streaming (issue #41 AC1).
 *
 * Wraps the model-gateway dispatch surface `POST /v1/agents/{agentId}/tasks`
 * (issue #16) behind typed, authenticated calls.  `dispatch` returns a typed
 * `TaskResult` whose `content` is the schema-validated typed object on a
 * served outcome and whose `outcome` is one of the closed set otherwise —
 * the SDK never fabricates content for a non-served outcome.  `stream` yields
 * the incremental dispatch events then the terminal `TaskResult`.
 */

import { TokenSource, verifyNotExpired } from "./auth.js";
import { ApiError, ConfigurationError } from "./errors.js";
import { errorFromEnvelope } from "./envelope.js";
import type { StreamTransport, Transport } from "./transport.js";
import { DispatchEvent, TaskRequest, TaskResult, TaskResultWire } from "./model.js";

/** Options for one task dispatch (the `POST .../tasks` body minus tenant). */
export interface DispatchOptions {
  input?: Record<string, unknown>;
  complexity?: number;
  tokens?: number;
  requestId?: string;
  metadata?: Record<string, unknown>;
}

export type GatewayStreamChunk = DispatchEvent | TaskResult;

export class GatewayClient {
  private readonly transport: Transport;
  private readonly tokenSource: TokenSource;
  private readonly tenantId?: string;

  constructor(transport: Transport, opts: { tokenSource?: TokenSource; tenantId?: string } = {}) {
    this.transport = transport;
    this.tokenSource = opts.tokenSource ?? new TokenSource();
    this.tenantId = opts.tenantId;
  }

  /** Resolve the current token + effective tenant (fail closed). */
  private tokenAndTenant(): { token: string; tenantId: string } {
    const token = this.tokenSource.require();
    const session = verifyNotExpired(token);
    const tenantId = this.tenantId ?? session.tenantId;
    if (!tenantId) {
      throw new ConfigurationError(
        "no tenant context: pass tenant_id or use a session token with a tenantId claim"
      );
    }
    return { token, tenantId };
  }

  /** Submit one task and return the typed `TaskResult`. */
  async dispatch(agentId: string, taskType: string, options: DispatchOptions = {}): Promise<TaskResult> {
    const { token, tenantId } = this.tokenAndTenant();
    const request: TaskRequest = {
      tenantId,
      taskType,
      input: options.input ?? {},
      complexity: options.complexity,
      tokens: options.tokens,
      stream: false,
      requestId: options.requestId,
      metadata: options.metadata,
    };
    const response = await this.transport.request("POST", `/v1/agents/${agentId}/tasks`, {
      body: request,
      token,
    });
    return parseGatewayResponse(response);
  }

  /** Dispatch and throw `TaskNotServedError` on a non-served outcome. */
  async run(agentId: string, taskType: string, options: DispatchOptions = {}): Promise<TaskResult> {
    return (await this.dispatch(agentId, taskType, options)).ensureServed();
  }

  /**
   * Stream one dispatch: yields incremental events, then the terminal result.
   * The transport must implement `StreamTransport` (SSE relay in production;
   * the offline doubles speak the same chunk contract).
   */
  async *stream(agentId: string, taskType: string, options: DispatchOptions = {}): AsyncGenerator<GatewayStreamChunk> {
    const transport = this.transport as unknown;
    const requestStream = (transport as StreamTransport).requestStream;
    if (typeof requestStream !== "function") {
      throw new ConfigurationError(
        "transport does not support streaming; provide a StreamTransport"
      );
    }
    const { token, tenantId } = this.tokenAndTenant();
    const request: TaskRequest = {
      tenantId,
      taskType,
      input: options.input ?? {},
      complexity: options.complexity,
      tokens: options.tokens,
      stream: true,
      requestId: options.requestId,
      metadata: options.metadata,
    };
    for await (const chunk of requestStream.call(transport, "POST", `/v1/agents/${agentId}/tasks`, {
      body: request,
      token,
    })) {
      if ("event" in chunk) {
        yield chunk.event as unknown as DispatchEvent;
      } else {
        yield parseGatewayResponse(chunk);
      }
    }
  }
}

/** Parse the issue #16 gateway envelope into a typed `TaskResult`. */
function parseGatewayResponse(response: Record<string, unknown>): TaskResult {
  if (typeof response !== "object" || response === null || !("result" in response)) {
    const envelope = response as unknown as {
      status?: number;
      error?: { code?: string; message?: string };
    };
    throw errorFromEnvelope({
      ok: false,
      status: envelope.status ?? 502,
      requestId: "",
      error: envelope.error ?? null,
    });
  }
  const raw = response.result as unknown as TaskResultWire;
  if (!raw.requestId && !raw.tenantId) {
    throw new ApiError(500, "malformed_response", "gateway envelope had no result payload");
  }
  return TaskResult.from(response.result as Record<string, unknown>);
}
